workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/CompositionRoot.kt

```diff
@@ -0,0 +1,112 @@
+ * [composition locals][androidx.compose.runtime.CompositionLocal] that all [composeViewFactory]
+ * factories need access to, such as UI themes.
+ *
+ * This function will called once, to wrap the _highest-level_ [composeViewFactory] in the tree.
+ * However, composition locals are propagated down to child [composeViewFactory] compositions, so
+ * any locals provided here will be available in _all_ [composeViewFactory] compositions.
+ */
+public typealias CompositionRoot = @Composable (content: @Composable () -> Unit) -> Unit
+
+/**
+ * Convenience function for applying a [CompositionRoot] to this [ViewEnvironment]'s [ViewRegistry].
+ * See [ViewRegistry.withCompositionRoot].
+ */
+@WorkflowUiExperimentalApi
+public fun ViewEnvironment.withCompositionRoot(root: CompositionRoot): ViewEnvironment =
+  this + (ViewRegistry to this[ViewRegistry].withCompositionRoot(root))
+
+/**
+ * Returns a [ViewRegistry] that ensures that any [composeViewFactory] factories registered in this
+ * registry will be wrapped exactly once with a [CompositionRoot] wrapper.
+ * See [CompositionRoot] for more information.
+ */
+@WorkflowUiExperimentalApi
+public fun ViewRegistry.withCompositionRoot(root: CompositionRoot): ViewRegistry =
+  mapFactories { factory ->
+    @Suppress("UNCHECKED_CAST")
+    (factory as? ComposeViewFactory<Any>)?.let { composeFactory ->
+      @Suppress("UNCHECKED_CAST")
+      composeViewFactory(composeFactory.type) { rendering, environment ->
+        WrappedWithRootIfNecessary(root) {
+          composeFactory.Content(rendering, environment)
+        }
+      }
+    } ?: factory
+  }
+
+/**
+ * Adds [content] to the composition, ensuring that [CompositionRoot] has been applied. Will only
+ * wrap the content at the highest occurrence of this function in the composition subtree.
+ */
+@VisibleForTesting(otherwise = PRIVATE)
+@Composable internal fun WrappedWithRootIfNecessary(
+  root: CompositionRoot,
+  content: @Composable () -> Unit
+) {
+  if (LocalHasViewFactoryRootBeenApplied.current) {
+    // The only way this local can have the value true is if, somewhere above this point in the
+    // composition, the else case below was hit and wrapped us in the local. Since the root
+    // wrapper will have already been applied, we can just compose content directly.
+    content()
+  } else {
+    // If the local is false, this is the first time this function has appeared in the composition
+    // so far. We provide a true value for the local for everything below us, so any recursive
+    // calls to this function will hit the if case above and not re-apply the wrapper.
+    CompositionLocalProvider(LocalHasViewFactoryRootBeenApplied provides true) {
+      root(content)
+    }
+  }
+}
+
+/**
+ * Applies [transform] to each [ViewFactory] in this registry. Transformations are applied lazily,
+ * at the time of lookup via [ViewRegistry.getFactoryFor].
+ */
+@WorkflowUiExperimentalApi
+private fun ViewRegistry.mapFactories(
```

Comment:
Wondering if this will bite us on our internal 0/1 glue. Also wondering if I care.

Reply:
Seems safe

Reply:
We might not even need it – since we have dedicated wrappers around `composeViewFactory` for our internal design system, we can use that as the bottleneck for introducing any special compose config.

Reply:
I am tempted to even not bother landing this, although some of the other samples and tests I think rely on it.